File: data_collection.r
# --------------------[ DEPENDENCIES ]--------------------
install.packages(c(
    "httr",
    "jsonlite",
    "progress"
))
library(httr)
library(jsonlite)
library(progress)

# --------------------[ CONFIGURATION ]--------------------
CONFIG <- list(
    api_key = readLines("config.txt")[1],
    networth_endpoint = "http://localhost:3000/networth",
    state_file = "state.json",
    output_file = "player_data.csv",
    uuid_file = "uuids.txt"
)

# --------------------[ STATE HANDLING ]--------------------
loadState <- function() {
    if (file.exists(CONFIG$state_file)) {
        state <- fromJSON(CONFIG$state_file)
        if (is.null(state$processed_uuids)) state$processed_uuids <- character()
        if (is.null(state$current_index)) state$current_index <- 1
        if (is.null(state$total)) state$total <- 0
        return(state)
    } else {
        return(list(
            processed_uuids = character(),
            current_index = 1,
            total = 0
        ))
    }
}

saveState <- function(processed_uuids, current_index, total) {
    state <- list(
        processed_uuids = processed_uuids,
        current_index = current_index,
        total = total
    )
    write(toJSON(state, auto_unbox = TRUE, pretty = TRUE), CONFIG$state_file)
}

# --------------------[ API FUNCTIONS ]--------------------
getProfileData <- function(uuid, retries = 3) {
    url <- sprintf("https://api.hypixel.net/v2/skyblock/profiles?uuid=%s&key=%s", uuid, CONFIG$api_key)
    for (i in 1:retries) {
        res <- GET(url)
        status <- status_code(res)

        if (status == 200) {
            data <- content(res, "parsed", type = "application/json")
            if (!isTRUE(data$success) || length(data$profiles) == 0) stop("No profile data")
            sel <- Filter(function(p) isTRUE(p$selected), data$profiles)
            if (length(sel) == 0) stop("No selected profile")
            p <- sel[[1]]
            return(list(
                profileData = p$members[[uuid]],
                bankBalance = if (!is.null(p$banking)) p$banking$balance else 0,
                profileId   = p$profile_id
            ))
        } else if (status == 429) {
            delay <- 2 * i
            cat(sprintf("⏳ Rate limited. Waiting %d seconds... (Attempt %d/%d)\n", delay, i, retries))
            Sys.sleep(delay)
        } else {
            stop("❌ API error ", status)
        }
    }
    return(NULL)
}

getMuseumData <- function(uuid, profileId) {
    url <- sprintf("https://api.hypixel.net/v2/skyblock/museum?profile=%s&key=%s", profileId, CONFIG$api_key)
    res <- GET(url)
    if (status_code(res) != 200) stop("Museum API failed (status ", status_code(res), ")")

    data <- content(res, "parsed", type = "application/json")
    if (!isTRUE(data$success) || !is.list(data$members)) stop("Museum data missing")
    if (!(uuid %in% names(data$members))) stop("No museum entry for UUID")

    data$members[[uuid]]
}

calculateNetworth <- function(profileData, museumData, bankBalance) {
    res <- POST(
        CONFIG$networth_endpoint,
        body = toJSON(list(
            profileData = profileData,
            museumData = museumData,
            bankBalance = bankBalance
        ), auto_unbox = TRUE),
        encode = "json",
        content_type_json()
    )

    if (status_code(res) != 200) stop("Networth API returned ", status_code(res), ": ", content(res, "text"))
    result <- content(res, "parsed", type = "application/json")

    raw_nw <- result$networth
    nw_num <- as.numeric(raw_nw)
    if (is.na(nw_num)) stop("Invalid networth value: ", raw_nw)

    formatC(nw_num, format = "f", digits = 2, big.mark = ",")
}

# --------------------[ DATA HELPERS ]--------------------
initializeDataFrame <- function() {
    if (file.exists(CONFIG$output_file)) {
        return(read.csv(CONFIG$output_file, stringsAsFactors = FALSE))
    } else {
        return(data.frame(
            uuid = character(),
            magical_power = numeric(),
            level = numeric(),
            networth = character(),
            stringsAsFactors = FALSE
        ))
    }
}

appendRecord <- function(df, uuid, mp, lvl, nw) {
    df[nrow(df) + 1, ] <- list(uuid = uuid, magical_power = mp, level = lvl, networth = nw)
    return(df)
}

safeExtract <- function(obj, path, default = NA) {
    tryCatch(
        {
            val <- eval(parse(text = paste0("obj$", path)))
            if (is.null(val)) default else val
        },
        error = function(e) default
    )
}

formatDuration <- function(seconds) {
    days <- seconds %/% 86400
    hours <- (seconds %% 86400) %/% 3600
    minutes <- (seconds %% 3600) %/% 60
    secs <- seconds %% 60
    sprintf("%dd %02dh %02dm %02ds", days, hours, minutes, secs)
}

# --------------------[ MAIN LOOP ]--------------------
collectDataToCSV <- function() {
    uuids <- unique(trimws(readLines(CONFIG$uuid_file)))
    uuids <- uuids[uuids != ""]

    state <- loadState()
    processed <- state$processed_uuids
    current_index <- state$current_index
    total <- if (state$total == 0) length(uuids) else state$total
    data_for_csv <- initializeDataFrame()

    start_time <- Sys.time()
    remaining <- length(uuids) - current_index + 1

    for (i in current_index:length(uuids)) {
        uuid <- uuids[i]
        elapsed <- as.numeric(Sys.time() - start_time, units = "secs")
        est_total <- (elapsed / (i - current_index + 1)) * remaining
        eta <- formatDuration(round(est_total))

        cat(sprintf("\n📦 Progress: %d / %d | ETA: %s\n", i, total, eta))

        if (uuid %in% processed) {
            next
        }

        saveState(processed, i, total)
        prof <- NULL

        tryCatch(
            {
                prof <- getProfileData(uuid)
            },
            error = function(e) {
                cat("❌ Profile error for", uuid, ":", conditionMessage(e), "\n")
            }
        )

        if (is.null(prof)) next

        tryCatch(
            {
                pd <- prof$profileData
                bb <- prof$bankBalance
                pid <- prof$profileId
                md <- getMuseumData(uuid, pid)

                mp <- safeExtract(pd, "accessory_bag_storage$highest_magical_power", 0)
                lvl <- safeExtract(pd, "leveling$experience", 0) / 100
                nw <- calculateNetworth(pd, md, bb)

                cat(sprintf("✅ %s — Level: %.2f | MP: %d | NW: %s\n", uuid, lvl, mp, nw))

                data_for_csv <- appendRecord(data_for_csv, uuid, mp, lvl, nw)
                write.csv(data_for_csv, CONFIG$output_file, row.names = FALSE)

                processed <- c(processed, uuid)
                saveState(processed, i + 1, total)
            },
            error = function(e) {
                cat("❌ Error processing", uuid, ":", conditionMessage(e), "\n")
            }
        )
    }

    cat("\n🎉 Data collection complete. Total entries:", nrow(data_for_csv), "\n")
    saveState(processed, 1, total)
}

# --------------------[ EXECUTION ]--------------------
collectDataToCSV()

File: data_collection.js
const fs = require("fs");
const { ProfileNetworthCalculator } = require("skyhelper-networth");

const apiKey = fs.readFileSync("config.txt", "utf8").trim();
const uuidFile = "uuids.txt";
const stateFile = "data_state.json";
const outputFile = "player_data.csv";

const MAX_THROTTLE_ERRORS = 10;
const THROTTLE_COOLDOWN_BASE = 5000;
let throttleErrors = 0;

function loadState() {
	if (fs.existsSync(stateFile)) {
		const s = JSON.parse(fs.readFileSync(stateFile));
		return {
			processed: s.processed || [],
			index: s.index || 0,
			total: s.total || null,
			lastThrottleTime: s.lastThrottleTime || 0,
		};
	}
	return { processed: [], index: 0, total: null, lastThrottleTime: 0 };
}

function saveState(processed, index, total, lastThrottleTime) {
	fs.writeFileSync(stateFile, JSON.stringify({ processed, index, total, lastThrottleTime }, null, 2));
}

async function fetchJSON(url) {
	const state = loadState();

	const timeSinceLastThrottle = Date.now() - state.lastThrottleTime;
	if (state.lastThrottleTime > 0 && timeSinceLastThrottle < THROTTLE_COOLDOWN_BASE) {
		const waitTime = THROTTLE_COOLDOWN_BASE - timeSinceLastThrottle;
		console.log(`⏳ Waiting ${waitTime}ms due to recent throttling...`);
		await new Promise((resolve) => setTimeout(resolve, waitTime));
	}

	for (let i = 1; i <= 3; i++) {
		try {
			const res = await fetch(url);

			if (res.status === 429) {
				throttleErrors++;
				console.warn(`⚠️ Rate limited (429). Retry ${i}/3... (Total: ${throttleErrors}/${MAX_THROTTLE_ERRORS})`);

				if (throttleErrors >= MAX_THROTTLE_ERRORS) {
					throw new Error("Maximum throttle errors reached. Stopping execution.");
				}

				const backoffTime = THROTTLE_COOLDOWN_BASE * Math.pow(2, i);
				console.log(`⏳ Backing off for ${backoffTime}ms...`);
				await new Promise((r) => setTimeout(r, backoffTime));
				continue;
			}

			if (!res.ok) throw new Error(`Fetch failed: ${res.status}`);
			const data = await res.json();

			if (data.throttle === true) {
				throttleErrors++;
				const throttleTime = Date.now();
				saveState(state.processed, state.index, state.total, throttleTime);

				console.warn(`⚠️ API throttled (${data.cause}). Retry ${i}/3... (Total: ${throttleErrors}/${MAX_THROTTLE_ERRORS})`);

				if (throttleErrors >= MAX_THROTTLE_ERRORS) {
					throw new Error("Maximum throttle errors reached. Stopping execution.");
				}

				const backoffTime = THROTTLE_COOLDOWN_BASE * Math.pow(2, i);
				console.log(`⏳ Backing off for ${backoffTime}ms...`);
				await new Promise((r) => setTimeout(r, backoffTime));
				continue;
			}

			return data;
		} catch (err) {
			if (i === 3 || (!err.message.includes("Fetch failed: 429") && !err.message.includes("Maximum throttle errors"))) {
				throw err;
			}

			if (err.message.includes("Maximum throttle errors")) {
				throw err;
			}
		}
	}
	throw new Error("Max retries reached");
}

async function getProfileData(uuid) {
	const url = `https://api.hypixel.net/v2/skyblock/profiles?uuid=${uuid}&key=${apiKey}`;
	try {
		const data = await fetchJSON(url);
		if (!data.success) {
			if (data.cause === "Key throttle" && data.throttle === true) {
				throw new Error("API throttled");
			}
			throw new Error(`API returned error: ${data.cause || "Unknown error"}`);
		}

		if (!data.profiles?.length) throw new Error("No profile data");
		const sel = data.profiles.find((p) => p.selected);
		if (!sel) throw new Error("No selected profile");
		return {
			profileData: sel.members[uuid],
			bankBalance: (sel.banking && sel.banking.balance) || 0,
			profileId: sel.profile_id,
		};
	} catch (err) {
		console.error(`❌ Profile error for ${uuid}: ${err.message}`);
		if (err.message.includes("Maximum throttle errors") || err.message.includes("API throttled")) {
			throw err;
		}
		return null;
	}
}

async function getMuseumData(uuid, profileId) {
	const url = `https://api.hypixel.net/v2/skyblock/museum?profile=${profileId}&key=${apiKey}`;
	try {
		const data = await fetchJSON(url);
		if (!data.success) {
			if (data.cause === "Key throttle" && data.throttle === true) {
				throw new Error("API throttled");
			}
			throw new Error(`API returned error: ${data.cause || "Unknown error"}`);
		}

		if (!data.members[uuid]) {
			throw new Error("Museum data missing");
		}
		return data.members[uuid];
	} catch (err) {
		if (err.message.includes("Maximum throttle errors") || err.message.includes("API throttled")) {
			throw err;
		}
		throw err;
	}
}

async function calculateNetworth(profileData, museumData, bankBalance) {
	const calculator = new ProfileNetworthCalculator(profileData, museumData, bankBalance);
	const result = await calculator.getNetworth();
	return result.networth;
}

function writeCsvRow(filePath, row, isFirstWrite = false) {
	const line = `"${row.uuid}",${row.magical_power},${row.level},"${row.networth}"\n`;
	if (isFirstWrite) {
		const header = "UUID,MagicalPower,Level,Networth\n";
		fs.writeFileSync(filePath, header + line, { encoding: "utf8" });
	} else {
		fs.appendFileSync(filePath, line, { encoding: "utf8" });
	}
}

async function main() {
	const uuids = fs
		.readFileSync(uuidFile, "utf8")
		.split("\n")
		.map((l) => l.trim())
		.filter(Boolean);

	let { processed, index, total, lastThrottleTime } = loadState();
	total = total || uuids.length;

	const isFirstWrite = !fs.existsSync(outputFile);

	console.log(`Starting processing at index ${index}/${total} (${processed.length} already processed)`);

	if (lastThrottleTime > 0) {
		const timeSinceLastThrottle = Date.now() - lastThrottleTime;
		if (timeSinceLastThrottle < THROTTLE_COOLDOWN_BASE * 2) {
			const waitTime = Math.max(0, THROTTLE_COOLDOWN_BASE * 2 - timeSinceLastThrottle);
			console.log(`⏳ Waiting ${waitTime}ms before starting due to previous throttling...`);
			await new Promise((resolve) => setTimeout(resolve, waitTime));
		}
	}

	try {
		for (let i = index; i < uuids.length; i++) {
			const uuid = uuids[i];
			if (processed.includes(uuid)) {
				continue;
			}

			const currentPosition = i + 1;
			const progressPercentage = ((currentPosition / total) * 100).toFixed(2);
			console.log(`📊 Processing ${currentPosition}/${total} (${progressPercentage}%) - UUID: ${uuid}`);

			saveState(processed, i, total, lastThrottleTime);

			if (i > index) {
				await new Promise((resolve) => setTimeout(resolve, 300));
			}

			const prof = await getProfileData(uuid);
			if (!prof) {
				continue;
			}

			let museum;
			try {
				museum = await getMuseumData(uuid, prof.profileId);
			} catch (e) {
				console.error(`❌ Museum error for ${uuid}: ${e.message}`);
				if (e.message.includes("Maximum throttle errors") || e.message.includes("API throttled")) {
					throw e;
				}
				continue;
			}

			try {
				const networth = await calculateNetworth(prof.profileData, museum, prof.bankBalance);
				const mp = prof.profileData.accessory_bag_storage?.highest_magical_power || 0;
				const lvl = (prof.profileData.leveling?.experience || 0) / 100;

				writeCsvRow(
					outputFile,
					{
						uuid,
						magical_power: mp,
						level: lvl.toFixed(2),
						networth: networth.toFixed(2),
					},
					isFirstWrite && i === index
				);

				console.log(`✅ ${uuid} — Level: ${lvl.toFixed(2)} | MP: ${mp} | NW: ${networth.toFixed(2)}`);
				processed.push(uuid);
                throttleErrors = 0;
			} catch (e) {
				console.error(`❌ Calculation error for ${uuid}: ${e.message}`);
				if (e.message.includes("Maximum throttle errors") || e.message.includes("API throttled")) {
					throw e;
				}
			}

			saveState(processed, i + 1, total, lastThrottleTime);
		}

		console.log(`\n🎉 Data collection complete: ${processed.length}/${total} (100%)`);
		saveState([], 0, total, 0);
	} catch (error) {
		if (error.message.includes("Maximum throttle errors") || error.message.includes("API throttled")) {
			console.log(`\n⛔ Stopping execution after ${MAX_THROTTLE_ERRORS} throttle errors.`);
			console.log(`Progress saved at index ${index}. Run the script again later.`);
		} else {
			console.error(`\n❌ Unexpected error: ${error.message}`);
		}
	}
}

main();
